### ISSUE TITLE
Picture element: Full size image is missing in PICTURE > SOURCE srcset

### ISSUE THREAD

**Issue by mukeshpanchal27** [WordPress/performance#1417]
## Bug Description

As discussed https://github.com/WordPress/performance/pull/1408#discussion_r1695593652, the full size image is missing in picture element source srcset.

## Steps to reproduce

1. Activate the Modern Image Formats plugin and enable picture support from the Media Settings screen.
2. Add an image block
3. Preview the page in the front end, Check the PICTURE > SOURCE srcset.

IMG srcset:
```
http://localhost:8888/wp-content/uploads/2024/07/car-1024x576.jpeg 1024w,
http://localhost:8888/wp-content/uploads/2024/07/car-300x169.jpeg 300w,
http://localhost:8888/wp-content/uploads/2024/07/car-768x432.jpeg 768w,
http://localhost:8888/wp-content/uploads/2024/07/car.jpeg 1080w
```

Picture > source srcset:
```
http://localhost:8888/wp-content/uploads/2024/07/car-1024x576.jpeg 1024w,
http://localhost:8888/wp-content/uploads/2024/07/car-300x169.jpeg 300w,
http://localhost:8888/wp-content/uploads/2024/07/car-768x432.jpeg 768w
```

## Screenshots

<img width="1576" alt="Screenshot 2024-07-30 at 4 38 57 PM" src="https://github.com/user-attachments/assets/68cc541e-35c3-4cc1-9f2e-f6145672847d">

**Comment by devansh016** [comment]
Hi @mukeshpanchal27,

In line 25, `wp_get_attachment_metadata` is not getting `sources` field for bigger sizes image ( greater than 1024 width).
Therefore, in line 34 `isset( $size['sources'] )` is false and `$mime_type_data` don't have sizes greater than 1024.

https://github.com/WordPress/performance/blob/08eba80fd8609add2df44d9960dc011868dc5d93/plugins/webp-uploads/picture-element.php#L25-L41

Sample Data for reference:
[mime_type_data](https://pastebin.com/ws5Am8AV)
[image_meta['sizes']](https://pastebin.com/wCphTD99).  1536x1536 , 2048x2048 don't have sources

**Comment by mukeshpanchal27** [comment]
@devansh016 Thanks for checking these.

The `wp_get_attachment_image_srcset()` function uses `wp_calculate_image_srcset()` to get the `srcset` for the image.
- The core uses some hacky way to add the full-size image in `srcset`.
- In https://github.com/WordPress/wordpress-develop/blob/6.6/src/wp-includes/media.php#L1339-L1349, the core added the full-size image to the post meta array.

We should add similar code to include the full-size image in the sub sizes array, like the core did. This should be done before the sizes `foreach` loop. Unit tests need to be added as well.

```php
$image_sizes = $image_meta['sizes'];

/*
 * WordPress flattens animated GIFs into one frame when generating intermediate sizes.
 * To avoid hiding animation in user content, if src is a full size GIF, a srcset attribute is not generated.
 * If src is an intermediate size GIF, the full size is excluded from srcset to keep a flattened GIF from becoming animated.
 */
if ( ! isset( $image_sizes['thumbnail']['mime-type'] ) || 'image/gif' !== $image_sizes['thumbnail']['mime-type'] ) {
	$image_full_size = array(
		'width'   => $image_meta['width'],
		'height'  => $image_meta['height'],
		'sources' => $image_meta['sources'],
	);
	array_unshift( $image_sizes, $image_full_size );
}
```

@devansh016, do you want to take ownership of this and see it through to the finish line?

cc. @adamsilverstein @westonruter @joemcgill

**Comment by devansh016** [comment]
@mukeshpanchal27 Yes, I would like to take ownership of this issue.

I found the actual problem `$sizes_with_mime_type_support` don't have image size like 1536x1536 and 2048x2048 etc. and hence some sizes doesn't have `sources` field. This problem only occur for images which are higher in resolution. 
https://github.com/WordPress/performance/blob/08eba80fd8609add2df44d9960dc011868dc5d93/plugins/webp-uploads/hooks.php#L173

Should we update the `$allowed_sizes` to include those?

https://github.com/WordPress/performance/blob/08eba80fd8609add2df44d9960dc011868dc5d93/plugins/webp-uploads/hooks.php#L709-L717

Can you suggest what approach should we take to solve this?

**Comment by 1ucay** [comment]
Hi, I have problem with similar, I need picture element with fallback jpg, but if I upload smaller image then medium size ( 768px ), picture element not populated, there is only avif in img src. Thank you

    <?php echo apply_filters( 'wp_content_img_tag', get_the_post_thumbnail( null, 'medium' ), 'the_content', get_post_thumbnail_id() ); ?>

**Comment by mukeshpanchal27** [comment]
@1ucay Please share Output of the function.

**Comment by 1ucay** [comment]
Image itself is 600 × 400 in jpg, in admin there are thumbnail (150x150px and full size ). No medium,no medium_large etc, because medium is 768x512px with center crop.

![scrrr](https://github.com/user-attachments/assets/db53e1ee-d36f-43f8-9e1c-3c4e16e0c5bb)

    <picture class="wp-picture-1157" style="display: contents">
        <img
            decoding="async"
            width="600"
            height="400"
            src="https://REDACTED/2024/04/veletrh-ptacek-jpg.avif"
            class="attachment-medium size-medium not-transparent"
            alt="veletrh ptacek"
            loading="lazy"
            data-has-transparency="false"
            data-dominant-color="73838b"
            style="--dominant-color: #73838b" />
    </picture>



function webp_uploads_wrap_image_in_picture()

    $image_meta              = wp_get_attachment_metadata( $attachment_id );
    print_r( $image_meta ); die();


    Array
    (
        [width] => 600
        [height] => 400
        [file] => 2024/04/veletrh-ptacek.jpg
        [filesize] => 30756
        [sizes] => Array
            (
                [thumbnail] => Array
                    (
                        [file] => veletrh-ptacek-150x150.jpg
                        [width] => 150
                        [height] => 150
                        [mime-type] => image/jpeg
                        [filesize] => 5646
                        [sources] => Array
                            (
                                [image/jpeg] => Array
                                    (
                                        [file] => veletrh-ptacek-150x150.jpg
                                        [filesize] => 5646
                                    )

                                [image/avif] => Array
                                    (
                                        [file] => veletrh-ptacek-150x150-jpg.avif
                                        [filesize] => 2615
                                    )

                            )

                    )

            )

        [image_meta] => Array
            (
                [aperture] => 0
                [credit] => 
                [camera] => 
                [caption] => 
                [created_timestamp] => 0
                [copyright] => 
                [focal_length] => 0
                [iso] => 0
                [shutter_speed] => 0
                [title] => 
                [orientation] => 0
                [keywords] => Array
                    (
                    )

            )

        [dominant_color] => 73838b
        [has_transparency] => 
        [sources] => Array
            (
                [image/jpeg] => Array
                    (
                        [file] => veletrh-ptacek.jpg
                        [filesize] => 30756
                    )

                [image/avif] => Array
                    (
                        [file] => veletrh-ptacek-jpg.avif
                        [filesize] => 16325
                    )

            )

    )

**Comment by mukeshpanchal27** [comment]
@1ucay Thanks for the details.

Could you please open a separate issue as this is different from the current one?

The WP core doesn't return `srcset` for the thumbnail size image, so the [picture element skips adding](https://github.com/WordPress/performance/blob/trunk/plugins/webp-uploads/picture-element.php#L121-L123) the `source` attribute. This is the expected result. For the fallback JPEG version, we already discussed it here: https://github.com/WordPress/performance/pull/1408#issuecomment-2268302619.

In my opinion, the result should be an IMG tag without a picture element if the `source` is null.

I did like to get @westonruter and @adamsilverstein thoughts here.

**Comment by 1ucay** [comment]
Thank you. So if I upload 150x150px or 500x200px (medium size is 768px) image, there is no way to support avif and jpg in picture element? Maybe we can have some htaccess fallback, when user browser dont support AVIF? For me it is more important serve both format in picture element, then srcset attribute. Thank you

**Comment by mukeshpanchal27** [comment]
As mention in previous reply. Check conversion in other issue for the fallback image.

> The WP core doesn't return `srcset` for the thumbnail size image, so the [picture element skips adding](https://github.com/WordPress/performance/blob/trunk/plugins/webp-uploads/picture-element.php#L121-L123) the `source` attribute. This is the expected result. For the fallback JPEG version, we already discussed it here: [#1408 (comment)](https://github.com/WordPress/performance/pull/1408#issuecomment-2268302619).

**Comment by mukeshpanchal27** [comment]
> Can you suggest what approach should we take to solve this?

@devansh016 Go with approach i suggest here: https://github.com/WordPress/performance/issues/1417#issuecomment-2262824727. We can always add those additional sizes through filter. The function only account for some image sizes and added filter if anyone want to extent to add additional size.

```php
add_filter( 'webp_uploads_image_sizes_with_additional_mime_type_support', static function ( $allowed_sizes ): array {
	$allowed_sizes['1536x1536'] = true;
	$allowed_sizes['2048x2048'] = true;
	return $allowed_sizes;
});
```

**Comment by westonruter** [comment]
As I just mentioned in https://github.com/WordPress/performance/pull/1430#issuecomment-2270026509:

> Why isn't there a fallback JPEG for this generated AVIF image? The size of the image being uploaded shouldn't matter. If uploading a 10x10 JPEG image, this should still result in an AVIF/WebP image being generated in addition to the JPEG fallback (if that option is enabled). The `PICTURE` should still be used even if the `IMG` lacks a `srcset` because it should provide the JPEG fallback for the AVIF.

**Comment by mukeshpanchal27** [comment]
@devansh016 I'm going to take over this as we have a release in the upcoming weeks. We need to mark this ready for review and merge.

**Comment by adamsilverstein** [comment]
> > Can you suggest what approach should we take to solve this?
> 
> @devansh016 Go with approach i suggest here: [#1417 (comment)](https://github.com/WordPress/performance/issues/1417#issuecomment-2262824727). We can always add those additional sizes through filter. The function only account for some image sizes and added filter if anyone want to extent to add additional size.
> 
> ```
> add_filter( 'webp_uploads_image_sizes_with_additional_mime_type_support', static function ( $allowed_sizes ): array {
> 	$allowed_sizes['1536x1536'] = true;
> 	$allowed_sizes['2048x2048'] = true;
> 	return $allowed_sizes;
> });
> ```

can we add this by default, unless a plugin has explicitly removed the core hooked sizes (`add_action( 'plugins_loaded', '_wp_add_additional_image_sizes', 0 )`) - that way, the sizes we generate will match the jpeg sizes core generates which is what we want. Once we are generating all of the sizes correctly, they should show up in the srcset entries.
